MassSpectrometry/MsDataFile.cs

```diff
@@ -83,63 +83,175 @@ public static int TopNpeakHelper(ref double[] intensities, ref double[] mArray,
+        {
+            Array.Sort(mArray, intensities);
+            int numberOfWindows = (int)Math.Round((scanRangeMaxMz - scanRangeMinMz + discreteMassBin) / discreteMassBin, 0);
+
+            FilteringParams firstFilter = new FilteringParams(1, percentMaxThreshold / 100, numberOfWindows, false, false);
 
-            // this loop breaks a scan up into "windows" (e.g., a scan with 100 peaks
-            // divided into 10 windows would have 10 peaks per window) and takes the top N peaks per window.
-            // the results of each trimmed window are concatenated into mzResults and intensityResults
-            for (int i = 0; i < filteringParams.NumberOfWindows; i++)
+            for (int i = 0; i < mArray.Length; i++)
             {
-                // make the last window end at the end of the spectrum
-                // this is to prevent rounding errors in windowSize from cutting off the end of the spectrum
-                if (i == filteringParams.NumberOfWindows - 1)
+                if (intensities[i] != 0)
                 {
-                    windowPeakIndexMaximum = intensities.Length - 1;
+                    intensities[i] = Math.Sqrt(intensities[i]);
                 }
+            }
 
-                // avoid index out of range problems
-                if (windowPeakIndexMinimum >= intensities.Length)
+            //set intensity of precursor ions to zero
+            for (int i = 0; i < mArray.Length; i++)
+            {
+                if (mArray[i] > (precursorMz - precursorDiscardRange) && mArray[i] < (precursorMz + precursorDiscardRange))
                 {
-                    break;
+                    intensities[i] = 0;
                 }
+            }
 
-                // determine the valid peaks given filtering conditions for this window
-                var windowIntensities = new double[windowPeakIndexMaximum - windowPeakIndexMinimum + 1];
-                Array.Copy(intensities, windowPeakIndexMinimum, windowIntensities, 0, windowIntensities.Length);
+            //filter peaks into discrete mass bins
+            MsDataFile.WindowModeHelper(ref intensities, ref mArray, firstFilter, (scanRangeMinMz - discreteMassBin / 2), (scanRangeMaxMz + discreteMassBin / 2));
 
-                var windowMzs = new double[windowPeakIndexMaximum - windowPeakIndexMinimum + 1];
-                Array.Copy(mArray, windowPeakIndexMinimum, windowMzs, 0, windowMzs.Length);
+            //normalize intensity in segments to max 50
+            FilteringParams secondFilter = new FilteringParams((int)Math.Ceiling((decimal)numberOfWindows / 10), null, 10, false, false);
+            MsDataFile.WindowModeHelper(ref intensities, ref mArray, secondFilter, (scanRangeMinMz - discreteMassBin / 2), (scanRangeMaxMz + discreteMassBin / 2), 50);
```

Comment:
Why are we doing this multiple times?

Reply:
code efficiency

Reply:
I still don't understand. Can you add some comments for what's going on here? It seems like we're binning twice, which doesn't make sense to me.